metals/src/main/scala/scala/meta/internal/metals/CompilerConfiguration.scala

```diff
@@ -0,0 +1,344 @@
+
+  private val plugins = new CompilerPlugins()
+
+  sealed trait MtagsPresentationCompiler {
+    def await: PresentationCompiler
+  }
+
+  case class StandaloneCompiler(
+      scalaVersion: String,
+      symbolSearch: SymbolSearch,
+      classpath: Seq[Path],
+  ) extends MtagsPresentationCompiler {
+    private val mtags =
+      mtagsResolver.resolve(scalaVersion).getOrElse(MtagsBinaries.BuildIn)
+
+    val standalone: PresentationCompiler =
+      fromMtags(
+        mtags,
+        options = Nil,
+        classpath ++ Embedded.scalaLibrary(scalaVersion),
+        "default",
+        symbolSearch,
+      )
+
+    def await: PresentationCompiler = standalone
+  }
+
+  object StandaloneCompiler {
+
+    def apply(
+        scalaVersion: String,
+        classpath: Seq[Path],
+        sources: Seq[Path],
+        workspaceFallback: Option[SymbolSearch],
+    ): StandaloneCompiler = {
+      val search =
+        createStandaloneSearch(classpath, sources, workspaceFallback)
+      StandaloneCompiler(
+        scalaVersion,
+        search,
+        classpath,
+      )
+    }
+  }
+
+  trait LazyCompiler extends MtagsPresentationCompiler {
+
+    def buildTargetId: BuildTargetIdentifier
+    protected def fallback: PresentationCompiler
+    protected def newCompiler(classpath: Seq[Path]): PresentationCompiler
+
+    protected val presentationCompilerRef =
+      new ju.concurrent.atomic.AtomicReference[PresentationCompiler]()
+
+    private val wasResolved = new ju.concurrent.atomic.AtomicBoolean(false)
+    protected val presentationCompilerFuture: Future[PresentationCompiler] =
+      buildTargets
+        .targetClasspath(buildTargetId)
+        .getOrElse(Future.successful(Nil))
+        .map { classpath =>
+          // set wasResolved to avoid races on timeout below
+          wasResolved.set(true)
+          // Request finished, we can remove and shut down the fallback
+          Option(presentationCompilerRef.getAndSet(null)).foreach(_.shutdown())
+          val classpathSeq = classpath.toAbsoluteClasspath.map(_.toNIO).toSeq
+          newCompiler(classpathSeq)
+        }
+
+    def await: PresentationCompiler = {
+      val compilerConfig = config.initialConfig.compilers
+      try {
+        val pc = presentationCompilerRef.get()
+        if (pc != null) {
+          pc
+        } else {
+          val result = Await.result(
+            presentationCompilerFuture,
+            Duration(compilerConfig.timeoutDelay, compilerConfig.timeoutUnit),
+          )
+          presentationCompilerRef.set(result)
+          result
+        }
+      } catch {
+        case _: ju.concurrent.TimeoutException =>
+          scribe.warn(
+            s"Still waiting for information about classpath, using standalone compiler for now"
+          )
+          val pc = presentationCompilerRef.updateAndGet { old =>
+            if (old != null) old
+            else if (!wasResolved.get()) fallback
```

Comment:
I don't think creating standalone compiler has side effects? We don't create anything aside from the compiler. Do you think this could be an issue? We can use semaphores etc.

Reply:
Changed to synchronized and simplified a bit, so that we don't need the additional AtomicBoolean